### DISCUSSION TITLE
Feat(plugins/request-defer)

### DISCUSSION THREAD

**Discussion by WALL-E** [Kong/kong#7389] [Ideas and feature requests]
Recently I was developing a software firewall,  I need a plugin to delay all http requests when the system is in an abnormal state,  not limited or rejected, especially when doing **“flash deal”**

Two reasons
- Protect back-end services
- To minimize the losses，when there are a lot of robots to place  order

Others?

**Comment by Tieske** [comment]
have you seen #234?

**Comment by WALL-E** [comment]
@Tieske there could be significant difference,  for example, plug-in doesn't have enough ability to judge the requestor is a robot or not, it needs to rely on a third-party, risk management  or big-data.

Throttling algorithm seems rather coarsely.

**Comment by sonicaghi** [comment]
@WALL-E https://github.com/Mashape/kong/issues/154 ?

**Comment by WALL-E** [comment]
@sinzone no.  

usually, the faster the server side is rejected, the faster the client(robot)  request is sent,  the result is that server load may be getting higher and higher.

This kind of situation often appears in the mainland of China,  [京东](http://jd.com) or [淘宝](http://taobao.com), similar to Amazon

**Comment by WALL-E** [comment]
the plugin's schema.lua may be like

```
return {
  fields = {
    deferred_execution_ millisecond = { default = 100, type = "number" }
  }
}
```

**Comment by Tieske** [comment]
Sorry, not getting the requested behaviour just yet.

> I need a plugin to delay all http requests when the system is in an abnormal state, not limited or rejected,

Questions;
- so the plugin should defer **ALL** requests?
- when should it start returning errors?
- so you intent to use it as a devops tool? some monitoring tool triggers it and then you insert this plugin automatically through the management api? (and remove again when the situation normalizes)

> judge the requestor is a robot or not, it needs to rely on a third-party, risk management or big-data.

how does this fit into the way the plugin should work?

**Comment by WALL-E** [comment]
> so the plugin should defer ALL requests? 

Sorry, I may be a misnomer.  Not all, only to meet certain conditions of the request

> when should it start returning errors?

most of the time, do not return an error, just slow down the request

> so you intent to use it as a devops tool? some monitoring tool triggers it and then you insert this plugin automatically through the management api? (and remove again when the situation normalizes)

no devops tool, but user behavior analysis

may be need two config option (For each user)
- deferred execution millisecond
- expired time (auto delete or disable)

There are two roles
- decision maker: **user behavior analysis**
- executors: **kong/pugin-defer**
